File: server/scanner.py
"""Scan Gradle projects for Paparazzi screenshot failures.

Finds modules with build/paparazzi/failures/, detects current vs stale
failures using mtime clustering, matches golden images, and parses
JUnit XML for test statistics.
"""

import os
import re
import xml.etree.ElementTree as ET
from pathlib import Path

from server.filename_parser import parse_filename

MTIME_CLUSTER_TOLERANCE = 60.0

PRUNE_DIRS = {
    ".git",
    ".gradle",
    ".idea",
    "node_modules",
    ".cxx",
    ".transforms",
    "src",
    ".kotlin",
}


def scan_project(project_path):
    """Scan a Gradle project for Paparazzi failures. Blocking convenience wrapper."""
    result = {"modules": [], "failures": []}
    for phase, data in scan_project_incremental(project_path):
        if phase == "complete":
            result = data
    return result


def process_single_module(failures_dir, module_name, module_path, profiles=None):
    """Process a single Paparazzi module. Returns (module_data, failures_list).

    If profiles is provided, processes each profile's failure/golden dirs.
    Otherwise uses the legacy failures_dir with default golden dir.
    Reused by both the initial scan and the realtime watcher.
    """
    test_stats, xml_mtime = _parse_junit_xml(module_path)
    diff_pcts = _parse_diff_percentages(module_path)
    module_failures = []
    profile_counts = {}
    total_snapshots = 0

    if profiles:
        for profile in profiles:
            pname = profile["name"]
            f_dir = module_path / profile["failures_dir"]
            gp = _build_golden_patterns(profile)
            pf = _process_profile(
                f_dir,
                module_path,
                module_name,
                pname,
                test_stats,
                xml_mtime,
                golden_patterns=gp,
                diff_pcts=diff_pcts,
                delta_prefix=profile.get("delta_prefix", "delta-"),
                delta_suffix=profile.get("delta_suffix", ""),
                actual_suffix=profile.get("actual_suffix", ""),
            )
            module_failures.extend(pf)
            profile_counts[pname] = len(pf)
            g_dir_str = profile.get("golden_dir", "")
            if g_dir_str:
                g_dir = module_path / g_dir_str
                total_snapshots += (
                    len(list(g_dir.glob("*.png"))) if g_dir.is_dir() else 0
                )
    else:
        default_gp = ["src/test/snapshots/images/{name}.png"]
        pf = _process_profile(
            Path(failures_dir) if failures_dir else None,
            module_path,
            module_name,
            "baseline",
            test_stats,
            xml_mtime,
            golden_patterns=default_gp,
            diff_pcts=diff_pcts,
        )
        module_failures.extend(pf)
        profile_counts["baseline"] = len(pf)
        golden_dir = module_path / "src" / "test" / "snapshots" / "images"
        total_snapshots = (
            len(list(golden_dir.glob("*.png"))) if golden_dir.is_dir() else 0
        )

    module_data = {
        "name": module_name,
        "failures_path": str(failures_dir) if failures_dir else None,
        "golden_path": str(module_path / "src" / "test" / "snapshots" / "images"),
        "failure_count": len(module_failures),
        "snapshot_count": total_snapshots,
        "profile_counts": profile_counts,
        "test_stats": test_stats,
    }
    return module_data, module_failures


def _build_golden_patterns(profile):
    """Build golden pattern list from a profile config.

    If profile has 'golden_patterns', use those directly.
    Otherwise build from 'golden_dir' + optional 'golden_suffix'.
    Patterns use {name} as placeholder for snapshot name (without .png).
    """
    if "golden_patterns" in profile:
        return profile["golden_patterns"]
    g_dir = profile.get("golden_dir", "")
    suffix = profile.get("golden_suffix", "")
    patterns = []
    if suffix:
        patterns.append(f"{g_dir}/{{name}}{suffix}.png")
    patterns.append(f"{g_dir}/{{name}}.png")
    return patterns


def _resolve_golden(module_path, golden_patterns, snapshot_name):
    """Try each golden pattern in order, return first existing path or None.

    Patterns use {name} as placeholder for the snapshot name (without .png).
    Supports ** for recursive directory matching (glob).
    """
    name = snapshot_name[:-4] if snapshot_name.endswith(".png") else snapshot_name
    for pattern in golden_patterns:
        resolved = pattern.replace("{name}", name)
        if "**" in resolved:
            matches = list(module_path.glob(resolved))
            if matches:
                return matches[0]
        else:
            candidate = module_path / resolved
            if candidate.is_file():
                return candidate
    return None


def _process_profile(
    failures_dir,
    module_path,
    module_name,
    profile_name,
    test_stats,
    xml_mtime,
    golden_patterns=None,
    diff_pcts=None,
    delta_prefix="delta-",
    delta_suffix="",
    actual_suffix="",
):
    """Process one profile's failures for a module. Returns list of failure dicts."""
    if not golden_patterns:
        golden_patterns = []
    results = []
    if failures_dir and Path(failures_dir).is_dir():
        failures_dir = Path(failures_dir)
        current = _detect_current_failures(failures_dir, delta_prefix, delta_suffix)
        if xml_mtime > 0 and current and test_stats and test_stats["failed"] == 0:
            current = [f for f in current if f.stat().st_mtime > xml_mtime]

        # Resolve goldens once per failure (cache to avoid repeated ** globs)
        golden_cache = {}
        for f in current:
            base = _delta_to_base(f.name, delta_prefix, delta_suffix)
            golden_cache[f.name] = _resolve_golden(module_path, golden_patterns, base)

        for delta_path in current:
            base = _delta_to_base(delta_path.name, delta_prefix, delta_suffix)
            parsed = parse_filename(base)
            if actual_suffix:
                stem = base[:-4] if base.endswith(".png") else base
                actual_name = f"{stem}{actual_suffix}.png"
            else:
                actual_name = base
            actual_path = failures_dir / actual_name
            golden_path = golden_cache.get(delta_path.name)

            results.append(
                {
                    "module": module_name,
                    "profile": profile_name,
                    "filename": base,
                    "delta_path": str(delta_path),
                    "actual_path": str(actual_path) if actual_path.is_file() else None,
                    "golden_path": str(golden_path) if golden_path else None,
                    "package": parsed["package"],
                    "class_name": parsed["class_name"],
                    "method": parsed["method"],
                    "snapshot_name": parsed["snapshot_name"],
                    "status": "pending",
                    "diff_pct": (diff_pcts or {}).get(base),
                    "has_golden": golden_path is not None,
                    "has_actual": actual_path.is_file(),
                    "mtime": delta_path.stat().st_mtime,
                }
            )
    return results


def _delta_to_base(filename, delta_prefix, delta_suffix):
    """Convert a delta filename to the base snapshot name."""
    name = filename
    if delta_prefix and name.startswith(delta_prefix):
        name = name[len(delta_prefix) :]
    if delta_suffix:
        stem = name[:-4] if name.endswith(".png") else name
        if stem.endswith(delta_suffix):
            stem = stem[: -len(delta_suffix)]
            name = stem + ".png"
    return name


def scan_project_incremental(project_path, cancel_event=None, profiles=None):
    """Yield (phase, data) tuples as the scan progresses.

    Phases:
        ("discovered", {"total": N})
        ("progress", {"scanned": i, "current_module": name, "failures_so_far": N})
        ("cancelled", {})
        ("complete", {"modules": [...], "failures": [...]})
    """
    root = Path(project_path)

    # Phase 1: Fast discovery with os.walk + pruning
    # Discover modules that have build/paparazzi/ OR any profile's failures dir
    discovered = []
    for module_info in _discover_paparazzi_modules(root, profiles):
        discovered.append(module_info)
        yield ("discovering", {"found": len(discovered), "current_dir": module_info[1]})
    yield ("discovered", {"total": len(discovered)})

    # Phase 2: Process each module
    modules = []
    failures = []
    for i, (failures_dir, module_name, module_path) in enumerate(discovered):
        if cancel_event and cancel_event.is_set():
            yield ("cancelled", {})
            return

        module_data, module_failures = process_single_module(
            failures_dir, module_name, module_path, profiles
        )
        modules.append(module_data)
        if module_failures:
            failures.extend(module_failures)

        yield (
            "progress",
            {
                "scanned": i + 1,
                "current_module": module_name,
                "failures_so_far": len(failures),
            },
        )

    yield ("complete", {"modules": modules, "failures": failures})


def _discover_paparazzi_modules(root, profiles=None):
    """Walk project tree with aggressive pruning, yielding modules as found.

    Discovers any module with build/paparazzi/ (including passing modules).
    Yields (failures_dir_or_None, module_name, module_path) tuples.
    """
    root_str = str(root)

    for dirpath, dirnames, _filenames in os.walk(root_str):
        dirnames[:] = [d for d in dirnames if d not in PRUNE_DIRS]

        rel = os.path.relpath(dirpath, root_str)
        parts = rel.split(os.sep)

        # Aggressive pruning inside build/ directories
        if "build" in parts:
            idx = parts.index("build")
            depth = len(parts) - idx
            if depth == 1:
                dirnames[:] = [
                    d for d in dirnames if d in ("paparazzi", "test-results", "outputs")
                ]
            elif depth == 2 and parts[-1] == "outputs":
                # Inside build/outputs/ — only keep roborazzi
                dirnames[:] = [d for d in dirnames if d == "roborazzi"]
            elif depth == 3 and parts[-2:] == ["outputs", "roborazzi"]:
                # Found a roborazzi module
                module_parts = parts[:idx]
                if module_parts and module_parts != ["."]:
                    module_name = ":" + ":".join(module_parts)
                    module_path = root / Path(*module_parts)
                else:
                    module_name = ":root"
                    module_path = root
                yield (
                    Path(dirpath) if Path(dirpath).is_dir() else None,
                    module_name,
                    module_path,
                )
                dirnames.clear()
            elif depth == 2 and parts[-1] == "paparazzi":
                # Found a paparazzi module
                module_parts = parts[:idx]
                if module_parts and module_parts != ["."]:
                    module_name = ":" + ":".join(module_parts)
                    module_path = root / Path(*module_parts)
                else:
                    module_name = ":root"
                    module_path = root
                failures_dir = Path(dirpath) / "failures"
                yield (
                    failures_dir if failures_dir.is_dir() else None,
                    module_name,
                    module_path,
                )
                dirnames.clear()
            elif depth == 2 and parts[-1] == "test-results":
                pass
            else:
                dirnames.clear()


def _detect_current_failures(failures_dir, delta_prefix="delta-", delta_suffix=""):
    """Return only delta/compare files from the most recent test run."""
    delta_files = []
    for f in failures_dir.iterdir():
        if not f.is_file() or not f.name.endswith(".png"):
            continue
        is_delta = False
        if delta_prefix and f.name.startswith(delta_prefix):
            is_delta = True
        if delta_suffix:
            stem = f.name[:-4]
            if stem.endswith(delta_suffix):
                is_delta = True
        if is_delta:
            delta_files.append((f, f.stat().st_mtime))

    if not delta_files:
        return []

    delta_files.sort(key=lambda x: x[1], reverse=True)
    latest_mtime = delta_files[0][1]
    current = []
    for f, mtime in delta_files:
        if latest_mtime - mtime <= MTIME_CLUSTER_TOLERANCE:
            current.append(f)
        else:
            break

    return current


def _parse_diff_percentages(module_path):
    """Extract diff percentages from JUnit XML failure messages.

    Returns dict mapping delta filename to percentage (e.g., {"test.png": 0.044794}).
    Parses 'Images differ (by X.XXX%)' from failure message attributes.
    """
    result = {}
    for variant in ("testDebugUnitTest", "testReleaseUnitTest"):
        results_dir = module_path / "build" / "test-results" / variant
        if not results_dir.is_dir():
            continue
        for xml_file in results_dir.glob("TEST-*.xml"):
            try:
                tree = ET.parse(xml_file)
                for tc in tree.getroot().iter("testcase"):
                    for fail in tc.iter("failure"):
                        msg = fail.get("message", "")
                        # Extract percentage
                        pct_match = re.search(r"differ \(by ([\d.]+)%\)", msg)
                        # Extract delta filename
                        delta_match = re.search(r"delta-([^\s]+\.png)", msg)
                        if pct_match and delta_match:
                            # Store without delta- prefix
                            fname = delta_match.group(1)
                            # Get just the base filename
                            fname = fname.split("/")[-1]
                            result[fname] = float(pct_match.group(1))
            except (ET.ParseError, OSError):
                pass
    return result


def _parse_junit_xml(module_path):
    """Parse JUnit XML test results for a module.

    Looks in build/test-results/testDebugUnitTest/ and testReleaseUnitTest/.
    Returns (stats_dict, newest_xml_mtime) or (None, 0) if no XML found.
    """
    stats = {
        "tests": 0,
        "passed": 0,
        "failed": 0,
        "skipped": 0,
        "errors": 0,
        "time": 0.0,
    }
    found = False
    newest_mtime = 0.0

    for variant in ("testDebugUnitTest", "testReleaseUnitTest"):
        results_dir = module_path / "build" / "test-results" / variant
        if not results_dir.is_dir():
            continue
        for xml_file in results_dir.glob("TEST-*.xml"):
            parsed = _parse_single_junit_xml(xml_file)
            if parsed:
                found = True
                stats["tests"] += parsed["tests"]
                stats["failed"] += parsed["failed"]
                stats["skipped"] += parsed["skipped"]
                stats["errors"] += parsed["errors"]
                stats["time"] += parsed["time"]
                xml_mtime = xml_file.stat().st_mtime
                if xml_mtime > newest_mtime:
                    newest_mtime = xml_mtime

    if not found:
        return None, 0.0

    stats["passed"] = (
        stats["tests"] - stats["failed"] - stats["skipped"] - stats["errors"]
    )
    stats["time"] = round(stats["time"], 2)
    return stats, newest_mtime


def _parse_single_junit_xml(xml_path):
    """Parse a single JUnit XML file. Returns stats dict or None."""
    try:
        tree = ET.parse(xml_path)
        root = tree.getroot()
        return {
            "tests": int(root.get("tests", 0)),
            "failed": int(root.get("failures", 0)),
            "skipped": int(root.get("skipped", 0)),
            "errors": int(root.get("errors", 0)),
            "time": float(root.get("time", 0)),
        }
    except (ET.ParseError, ValueError, OSError):
        return None

File: src/scanner.js
/**
 * Scan a project for screenshot test failures.
 *
 * Module discovery and watch dirs are delegated to the strategy (e.g., gradle).
 * Detects current vs stale failures using mtime clustering,
 * matches golden images, and (for strategies that support JUnit) parses XML.
 */

const fs = require('fs');
const path = require('path');
const fg = require('fast-glob');
const { parseFilename } = require('./filenameParser');
const { parseJunitXml, parseDiffPercentages } = require('./junitParser');

const DEBUG = process.env.PAPASTUD_DEBUG === '1';
const log = (...args) => { if (DEBUG) console.log(...args); };

const MTIME_CLUSTER_TOLERANCE = 60.0;

const { getStrategy } = require('./strategies');
const { agp9Sibling } = require('./paparazziPaths');

function scanProject(projectPath, strategyName, opts = {}) {
  const result = { modules: [], failures: [] };
  for (const [phase, data] of scanProjectIncrementalSync(projectPath, null, null, strategyName, opts)) {
    if (phase === 'complete') return data;
  }
  return result;
}

function* scanProjectIncrementalSync(projectPath, cancelFn, profiles, strategyName, opts = {}) {
  const root = projectPath;
  const strategy = getStrategy(strategyName);

  // Strategy-driven path: one global parse (e.g. xcresult) returns failures
  // pre-bucketed by module, plus the module list — so we skip a second
  // discoverModules walk.
  let precomputed = null;
  if (typeof strategy.parseProjectFailures === 'function' && opts.cacheDir) {
    yield ['parsing', { source: 'strategy' }];
    precomputed = strategy.parseProjectFailures(root, opts.project || null, opts.cacheDir);
  }

  const moduleIter = precomputed && precomputed.modules
    ? precomputed.modules
    : strategy.discoverModules(root);

  const discovered = [];
  for (const moduleInfo of moduleIter) {
    discovered.push(moduleInfo);
    yield ['discovering', { found: discovered.length, current_dir: moduleInfo[1] }];
  }
  yield ['discovered', { total: discovered.length }];

  const modules = [];
  const failures = [];
  for (let i = 0; i < discovered.length; i++) {
    if (cancelFn && cancelFn()) {
      yield ['cancelled', {}];
      return;
    }

    const [failuresDir, moduleName, modulePath] = discovered[i];
    const [moduleData, moduleFailures] = precomputed
      ? processModuleFromPrecomputed(moduleName, modulePath, profiles, precomputed, failuresDir)
      : processSingleModule(failuresDir, moduleName, modulePath, profiles, strategyName);
    modules.push(moduleData);
    if (moduleFailures.length) failures.push(...moduleFailures);

    yield ['progress', {
      scanned: i + 1,
      current_module: moduleName,
      failures_so_far: failures.length,
    }];
  }

  yield ['complete', { modules, failures }];
}

function processModuleFromPrecomputed(moduleName, modulePath, profiles, precomputed, failuresDir) {
  const moduleFailures = precomputed.byModule.get(moduleName) || [];
  const profileCounts = {};
  for (const f of moduleFailures) {
    profileCounts[f.profile] = (profileCounts[f.profile] || 0) + 1;
  }

  let totalSnapshots = 0;
  let goldenPath = modulePath;
  if (profiles && profiles.length) {
    let primaryDirs = null;
    for (const p of profiles) {
      const dirs = effectiveGoldenDirs(modulePath, p.golden_dir);
      if (primaryDirs === null) primaryDirs = dirs;
      for (const d of dirs) totalSnapshots += countPngsRecursive(d);
    }
    if (primaryDirs && primaryDirs[0]) goldenPath = primaryDirs[0];
  }

  const moduleData = {
    name: moduleName,
    failures_path: failuresDir || null,
    golden_path: goldenPath,
    failure_count: moduleFailures.length,
    snapshot_count: totalSnapshots,
    profile_counts: profileCounts,
    test_stats: precomputed.stats,
  };
  return [moduleData, moduleFailures];
}

// On-disk golden directories worth inspecting for a module. Returns the
// configured directory plus its AGP-9 sibling when both exist, only the
// existing one when one exists, or `[configured]` as a fallback for
// empty/missing modules so callers using the first entry always get a
// reportable path.
function effectiveGoldenDirs(modulePath, goldenDir) {
  if (!goldenDir) return [];
  const primary = path.join(modulePath, goldenDir);
  const sibling = agp9Sibling(goldenDir);
  const agp9 = sibling ? path.join(modulePath, sibling) : null;
  const existing = [primary, agp9].filter(d => d && safeIsDir(d));
  return existing.length ? existing : [primary];
}

function safeIsDir(p) {
  try { return fs.statSync(p).isDirectory(); } catch { return false; }
}

// Sum the PNG counts across every effective golden dir for the module.
function totalGoldenPngs(modulePath, goldenDir) {
  return effectiveGoldenDirs(modulePath, goldenDir)
    .reduce((n, d) => n + countPngsRecursive(d), 0);
}

function processSingleModule(failuresDir, moduleName, modulePath, profiles, strategyName) {
  const strategy = getStrategy(strategyName);
  const useJunit = strategy.usesJunit !== false;
  const [testStats, xmlMtime] = useJunit ? parseJunitXml(modulePath) : [null, 0];
  const diffPcts = useJunit ? parseDiffPercentages(modulePath) : {};
  const moduleFailures = [];
  const profileCounts = {};
  let totalSnapshots = 0;
  let primaryGoldenDir = null;

  const hasProfiles = profiles && profiles.length;
  const goldenDirForReport = hasProfiles
    ? profiles[0].golden_dir
    : 'src/test/snapshots/images';

  if (hasProfiles) {
    for (const profile of profiles) {
      const pname = profile.name;
      const fDir = path.join(modulePath, profile.failures_dir);
      const gp = buildGoldenPatterns(profile);
      const pf = processProfile(
        fDir, modulePath, moduleName, pname,
        testStats, xmlMtime, gp, diffPcts,
        profile.delta_prefix !== undefined ? profile.delta_prefix : 'delta-',
        profile.delta_suffix || '',
        profile.actual_suffix || '',
      );
      moduleFailures.push(...pf);
      profileCounts[pname] = pf.length;
      const dirs = effectiveGoldenDirs(modulePath, profile.golden_dir);
      if (primaryGoldenDir === null) primaryGoldenDir = dirs[0] || null;
      for (const d of dirs) totalSnapshots += countPngsRecursive(d);
    }
  } else {
    const defaultGp = withAgp9Fallback(['src/test/snapshots/images/{name}.png']);
    const pf = processProfile(
      failuresDir, modulePath, moduleName, 'baseline',
      testStats, xmlMtime, defaultGp, diffPcts,
    );
    moduleFailures.push(...pf);
    profileCounts.baseline = pf.length;
    const dirs = effectiveGoldenDirs(modulePath, goldenDirForReport);
    primaryGoldenDir = dirs[0] || null;
    for (const d of dirs) totalSnapshots += countPngsRecursive(d);
  }

  const reportedGolden = primaryGoldenDir
    || path.join(modulePath, goldenDirForReport || 'src/test/snapshots/images');

  const moduleData = {
    name: moduleName,
    failures_path: failuresDir || null,
    golden_path: reportedGolden,
    failure_count: moduleFailures.length,
    snapshot_count: totalSnapshots,
    profile_counts: profileCounts,
    test_stats: testStats,
  };
  return [moduleData, moduleFailures];
}

function buildGoldenPatterns(profile) {
  if (profile.golden_patterns) return withAgp9Fallback([...profile.golden_patterns]);
  const gDir = profile.golden_dir || '';
  const suffix = profile.golden_suffix || '';
  const raw = [];
  if (suffix) raw.push(`${gDir}/{name}${suffix}.png`);
  raw.push(`${gDir}/{name}.png`);
  return withAgp9Fallback(raw);
}

// Pair every legacy Paparazzi pattern with its AGP-9 sibling so a profile
// persisted with only the old `src/test/snapshots/` path still resolves
// against a migrated module.
function withAgp9Fallback(patterns) {
  const out = [];
  const seen = new Set();
  for (const p of patterns) {
    if (!seen.has(p)) { out.push(p); seen.add(p); }
    const sibling = agp9Sibling(p);
    if (sibling && !seen.has(sibling)) { out.push(sibling); seen.add(sibling); }
  }
  return out;
}

// Candidate golden basenames to try, in preference order. Some tools (notably
// Roborazzi) URL-encode spaces in the emitted compare/delta filename while
// recording the golden with a literal space — e.g. delta
// `complete%20pause ad layout.….png` vs golden `complete pause ad layout.….png`.
// We try the name verbatim first, then variants that toggle `%20`↔space and a
// safely percent-decoded form, so the pair still matches either direction.
function goldenNameCandidates(name) {
  const out = [name];
  const add = (n) => { if (n && !out.includes(n)) out.push(n); };
  add(name.replace(/%20/gi, ' ')); // encoded space -> literal
  add(name.replace(/ /g, '%20'));  // literal space -> encoded
  try { add(decodeURIComponent(name)); } catch { /* malformed % — skip */ }
  return out;
}

function resolveGolden(modulePath, goldenPatterns, snapshotName) {
  const raw = snapshotName.endsWith('.png') ? snapshotName.slice(0, -4) : snapshotName;
  const names = goldenNameCandidates(raw);
  for (const pattern of goldenPatterns) {
    for (const name of names) {
      const resolved = pattern.replace('{name}', name);
      if (resolved.includes('**')) {
        const matches = fg.sync(resolved, { cwd: modulePath, absolute: true });
        if (matches.length) return matches[0];
      } else {
        const candidate = path.join(modulePath, resolved);
        try { if (fs.statSync(candidate).isFile()) return candidate; } catch {}
      }
    }
  }
  return null;
}

function processProfile(
  failuresDir, modulePath, moduleName, profileName,
  testStats, xmlMtime, goldenPatterns = [], diffPcts = {},
  deltaPrefix = 'delta-', deltaSuffix = '', actualSuffix = '',
) {
  const results = [];
  if (!failuresDir || !safeIsDir(failuresDir)) return results;

  let current = detectCurrentFailures(failuresDir, deltaPrefix, deltaSuffix);
  log(`[scan] ${profileName} in ${failuresDir}: ${current.length} candidates, xmlMtime=${xmlMtime}, testStats.failed=${testStats?.failed}`);
  // If JUnit XML reports 0 failures, all deltas are stale leftovers
  if (testStats && testStats.failed === 0 && xmlMtime > 0) {
    if (current.length) log(`[scan] all tests pass — filtering all ${current.length} stale deltas`);
    current = [];
  } else if (xmlMtime > 0 && current.length) {
    // Filter out delta files older than the latest JUnit XML run
    const before = current.length;
    current = current.filter(c => c.mtime > xmlMtime - MTIME_CLUSTER_TOLERANCE);
    if (current.length < before) log(`[scan] filtered ${before - current.length} stale deltas (older than xmlMtime ${xmlMtime})`);
  }

  const goldenCache = new Map();
  for (const c of current) {
    const base = deltaToBase(path.basename(c.path), deltaPrefix, deltaSuffix);
    // Key by full path so subdirs with same basename don't collide in compare mode
    goldenCache.set(c.path, resolveGolden(modulePath, goldenPatterns, base));
  }

  const noDeltaConvention = !deltaPrefix && !deltaSuffix;

  for (const c of current) {
    const candidatePath = c.path;
    const base = deltaToBase(path.basename(candidatePath), deltaPrefix, deltaSuffix);
    const parsed = parseFilename(base);
    const goldenPath = goldenCache.get(candidatePath);

    let actualPath, deltaFilePath;
    if (noDeltaConvention) {
      // Compare mode: the file IS the actual, there's no separate delta
      actualPath = candidatePath;
      deltaFilePath = null;
      // Skip if golden exists and files are identical (not a failure)
      if (goldenPath) {
        try {
          const aStat = fs.statSync(candidatePath);
          const gStat = fs.statSync(goldenPath);
          if (aStat.size === gStat.size) {
            const actualBuf = fs.readFileSync(candidatePath);
            const goldenBuf = fs.readFileSync(goldenPath);
            if (actualBuf.equals(goldenBuf)) continue;
          }
        } catch {}
      }
    } else {
      // Delta mode: separate delta file, actual may have a suffix.
      // Look for the actual alongside the delta — AGP 9 / KMP puts both in
      // `failures/androidMain/`, legacy puts both at the root of `failures/`.
      deltaFilePath = candidatePath;
      let actualName;
      if (actualSuffix) {
        const stem = base.endsWith('.png') ? base.slice(0, -4) : base;
        actualName = `${stem}${actualSuffix}.png`;
      } else {
        actualName = base;
      }
      actualPath = path.join(path.dirname(candidatePath), actualName);
      try { if (!fs.statSync(actualPath).isFile()) actualPath = null; }
      catch { actualPath = null; }
    }

    results.push({
      module: moduleName,
      profile: profileName,
      filename: base,
      delta_path: deltaFilePath,
      actual_path: actualPath,
      golden_path: goldenPath || null,
      package: parsed.package,
      class_name: parsed.class_name,
      method: parsed.method,
      snapshot_name: parsed.snapshot_name,
      status: 'pending',
      diff_pct: (diffPcts || {})[base] || null,
      has_golden: goldenPath !== null,
      has_actual: actualPath !== null,
      mtime: c.mtime,
    });
  }
  return results;
}

function deltaToBase(filename, deltaPrefix, deltaSuffix) {
  let name = filename;
  if (deltaPrefix && name.startsWith(deltaPrefix)) {
    name = name.slice(deltaPrefix.length);
  }
  if (deltaSuffix) {
    let stem = name.endsWith('.png') ? name.slice(0, -4) : name;
    if (stem.endsWith(deltaSuffix)) {
      stem = stem.slice(0, -deltaSuffix.length);
      name = stem + '.png';
    }
  }
  return name;
}

function detectCurrentFailures(failuresDir, deltaPrefix = 'delta-', deltaSuffix = '') {
  const deltaFiles = [];
  const noDeltaConvention = !deltaPrefix && !deltaSuffix;

  // Recurse into subdirs always: Paparazzi on AGP 9 / KMP writes failures under
  // `failures/androidMain/`, while the legacy layout puts them at the top of
  // `failures/`. A project can hold both side by side; mtime clustering then
  // picks the newest run.
  function scanDir(dir) {
    let entries;
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      if (entry.isDirectory()) {
        scanDir(path.join(dir, entry.name));
        continue;
      }
      if (!entry.isFile() || !entry.name.endsWith('.png')) continue;
      let isCandidate = noDeltaConvention;
      if (deltaPrefix && entry.name.startsWith(deltaPrefix)) isCandidate = true;
      if (deltaSuffix) {
        const stem = entry.name.slice(0, -4);
        if (stem.endsWith(deltaSuffix)) isCandidate = true;
      }
      if (isCandidate) {
        const fullPath = path.join(dir, entry.name);
        const mtime = fs.statSync(fullPath).mtimeMs / 1000;
        deltaFiles.push([fullPath, mtime]);
      }
    }
  }
  scanDir(failuresDir);

  if (!deltaFiles.length) return [];

  deltaFiles.sort((a, b) => b[1] - a[1]);
  const latestMtime = deltaFiles[0][1];
  const current = [];
  for (const [f, mtime] of deltaFiles) {
    if (latestMtime - mtime <= MTIME_CLUSTER_TOLERANCE) {
      current.push({ path: f, mtime });
    } else {
      break;
    }
  }
  return current;
}

// --- Compare mode (Compose Screenshot Testing) ---

function countPngsRecursive(dir) {
  let count = 0;
  try {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
      if (entry.isDirectory()) {
        count += countPngsRecursive(path.join(dir, entry.name));
      } else if (entry.name.endsWith('.png')) {
        count++;
      }
    }
  } catch {}
  return count;
}

module.exports = {
  scanProject,
  scanProjectIncrementalSync,
  processSingleModule,
  processModuleFromPrecomputed,
  deltaToBase,
  resolveGolden,
  buildGoldenPatterns,
  withAgp9Fallback,
  effectiveGoldenDirs,
  detectCurrentFailures,
};
